### ISSUE TITLE
[m1 support]Improve go build in dockerfile to support build docker images on m1 macbook

### ISSUE THREAD

**Issue by skyao** [dapr/dapr#4426]
## In what area(s)?

 /area test-and-release

## What version of Dapr?

master

## Expected Behavior

Develop on m1 macbook (mac + arm64), update code and build/push docker images to docker hub. Then run e2e tests and performance in k8s (linux + amd64) for test before commit the code to github.

The images should work on k8s (linux + amd64).

## Actual Behavior

The images for performance test don't work and the pod is CrashLoopBackOff with error "exec user process caused: exec format error":

```bash
$ k get pods -A
NAMESPACE     NAME                                     READY   STATUS             RESTARTS   AGE
dapr-tests    testapp-85d8d9db89-p6mms                 0/2     CrashLoopBackOff   13         11m

$ k logs testapp-85d8d9db89-p6mms -n dapr-tests testapp
standard_init_linux.go:228: exec user process caused: exec format error
```

## Steps to Reproduce the Problem

1. Build images on m1 macbook:

    ```bash
   $ make build-perf-app-all

   $ make push-perf-app-all
   ```

2. Run performance test case with remote k8s cluster

## Solution

The cause is that some test images is build by the dockerfile with `go build` command, but without GOOS and GOARCH settings. So the binary built on m1 book can't work on k8s ( linux + amd64).

To resolve this issue, we need to do two updates:

1. pass GOARCH as docker build argement in dapr makefile, such as dapr_tests.mk

    ```makefile
    define genPerfTestAppImageBuild
   .PHONY: build-perf-app-$(1)
   build-perf-app-$(1): check-e2e-env
	$(DOCKER) build -f $(PERF_TESTAPP_DIR)/$(1)/$(DOCKERFILE) $(PERF_TESTAPP_DIR)/$(1)/.
    -t    $(DAPR_TEST_REGISTRY)/perf-$(1):$(DAPR_TEST_TAG) 
   --build-arg GOARCH_ARG=$(GOARCH)
   endef
    ```
2. in the dockerfile with "go build", add GOOS and GOARCH settings

    ```makefile
    ARG GOARCH_ARG=amd64

    WORKDIR /app
    COPY app.go go.mod ./
    RUN go get -d -v && GOOS=linux GOARCH=$GOARCH_ARG go build -o app .
    ```

I verified for performance test cases and it works well.  The test cases of e2e test should do similar updates.

## Release Note

RELEASE NOTE:

**Comment by ItalyPaleAle** [comment]
I think this would only partially solve the issues, as no all test apps are written in Go. Some are in Java, .NET, Python, etc, and building those dockerfiles requires other steps besides building Go code.

A more complete solution would be to change how the E2E test apps are built and use Docker Buildx instead of the "regular" Docker Build. Then make sure that QEMU is installed if we need to cross-compile.

**Comment by skyao** [comment]
> I think this would only partially solve the issues, as no all test apps are written in Go. Some are in Java, .NET, Python, etc, and building those dockerfiles requires other steps besides building Go code.
> 
> A more complete solution would be to change how the E2E test apps are built and use Docker Buildx instead of the "regular" Docker Build. Then make sure that QEMU is installed if we need to cross-compile.

Agree with you to bring a complete solution. 

But before we achieve this target, can we do some small fix to improve the development experiences on m1 macbook? From my point, I use m1 macbook for daily work.

This PR is just fix somewhere, and I have verified only for perf test (including java and golang code to build docker images). Now I can code for the perf test on my m1 macbook and build the images without committing the code (yeah, at this time, the code is not verified and should not commit to github).

**Comment by ItalyPaleAle** [comment]
I think I'd rather just do it right first.

I spent a couple of hours this afternoon updating the dev container so it can work on ARM64. It's almost done and I just need to test a few things. That way, inside the dev container you'll be running ARM code natively and all apps will be built for ARM. Just give me a couple more days :) 

In the meanwhile, what I normally do in these cases when I need to X86, I use Codespaces!

**Comment by skyao** [comment]
> I think I'd rather just do it right first.
> 
> I spent a couple of hours this afternoon updating the dev container so it can work on ARM64. It's almost done and I just need to test a few things. That way, inside the dev container you'll be running ARM code natively and all apps will be built for ARM. Just give me a couple more days :)
> 
> In the meanwhile, what I normally do in these cases when I need to X86, I use Codespaces!

I have tried something similar on my m1 macbook. Docker and k8s (run in docker desktop) work well, but dapr need some arm64 images, for example redis for arm64. Another problem is that helm doesn't support to install redis arm64.

Waiting your update.

**Comment by skyao** [comment]
> I may have misunderstood your requirement. I thought you wanted to build Intel images while on M1 so you can deploy them to AKS? Instead you want to use K8s locally? That's not something I was targeting.
> 
> If there's no Helm chart for ARM64 for our dependencies, then we have some issues :)

I have tried two solutions to work with my m1 macbook, the background is that I want to coding on m1 macbook, and build images to do test without committing these code changes to github. I don't want to commit untest code to github at that time.

The first solution is to run k8s in m1: I tried with docker desktop, I almost succeed except that there's no arm64 redis images (perhaps also for mongodb, kafka, etc). And also helm is not support redis arm64, so even I can build a arm64 redis images by myself, it's also some troubles for us.

Then I turned to the second solution: building amd64 images in m1 macbook and run k8s in azure k8s. It worked well, I have used this solution for my perf test.

So, I think we can add this improvement into dapr so that at least we can coding on m1 and do test somewhere without committing the untest coding.
